### DISCUSSION TITLE
Retrieve the Anthropics API key using 'auth-source'.

### DISCUSSION THREAD

**Discussion by marcuskammer** [karthink/gptel#275] [Announcements]
I wrote a function to retrieve the API secret for anthropics from auth source. Maybe you want to include it into your library?!

```
(defun get-anthropic-key ()
  "Retrieve the Anthropics API key using 'auth-source'."
  (require 'auth-source)
  (let* ((auth-source-creation-prompts
          '((secret . "API key for %u@%h: ")))
         (found (nth 0 (auth-source-search
                        :max 1
                        :host "api.anthropic.com"
                        :user "apikey"
                        :require '(:secret)
                        :create t)))
         (secret (plist-get found :secret)))
    (if (functionp secret)
        (funcall secret)
      secret)))
```

**Comment by karthink** [comment]
@marcuskammer Thanks!

Note that auth-source support for _all_ backends is already provided by `gptel-api-key-from-auth-source`, which is the default value of `gptel-api-key`, so `get-anthropic-key` should not be required.

You can test this by setting the gptel backend to Anthropic, then running `(gptel-api-key-from-auth-source)`.  It should return the key.

**Comment by marcuskammer** [comment]
Thank you, didnt know that. Mind you add this as an example to the docs, please? The docs are a bit misleading as it says:

> The :key can be a function that returns the key (more secure).

So i was thinking i had to develop my own function.

**Reply by karthink** [reply]
Thank you for the feedback, I'll improve the README.

**Comment by adrianadermon** [comment]
I'm having trouble setting this up correctly. I have a `~/.authinfo` file that looks like this:
```
machine api.openai.com login apikey password ****
machine api.anthropic.com login apikey password ****
machine kagi.com login apikey password ****
```


 The following config works:
```
(use-package gptel
  :config
  (gptel-make-anthropic "Claude"
    :stream t                             
    :key (gptel-api-key-from-auth-source "api.anthropic.com")
    )
  (gptel-make-kagi "Kagi"      
    :key (gptel-api-key-from-auth-source "kagi.com")
    )
  :custom
  (gptel-use-curl nil)
  (gptel-default-mode 'org-mode)
  )
```

However, my impression from this discussion, and from looking at the package code, was that it should work without the `:key` lines. However, when I remove them, I'm asked for a user name and password. Have I misunderstood something?

**Reply by marcuskammer** [reply]
```
  (gptel-make-anthropic "Claude"
  :stream t
  :key #'gptel-api-key-from-auth-source)
```

**Reply by karthink** [reply]
You could also do

```elisp
(gptel-make-anthropic "Claude"
  :stream t
  :key gptel-api-key)
```

Since the default value of `gptel-api-key` is the `gptel-api-key-from-auth-source` function.

**Reply by adrianadermon** [reply]
That works - thanks for the help!

**Reply by adrianadermon** [reply]
Btw, would it make sense to set `gptel-api-key` as the default argument to `:key` in the `gptel-make-` functions?

**Reply by karthink** [reply]
> Btw, would it make sense to set gptel-api-key as the default argument to :key in the gptel-make- functions?

It will fail if they aren't using auth-source to manage their API keys.

**Reply by gavinhughes** [reply]
To avoid repeating  `:key gptel-api-key` in all configs, how about:

```
if keyword :key exist
   [current behavior]
else
  retrieve token from auth-source
```

**Reply by spudlyo** [reply]
After also struggling to understand the relationship between `gptel-api-key` and `gptel-api-key-from-auth-source` (and also writing my own function to read from auth-source) I added PR #715 in an attempt to clarify how this works in the README.
